File: sort/selection_sort/java/SelectionSort.java
public class SelectionSort {

	//selection sort algorithm
	public static int[] selection(int[] arr) {
		//minimum index variable
		int min = 0;
		//each pass will have 1 element sorted
		for (int i = 0; i < arr.length; i++) {
			//each pass will find the lowest unsorted element and bring it to the bottom
			for (int j = 0; j < arr.length; j++) {
				//if the element is greater than the minimum, switch their values
				if(arr[j]>arr[min]){
					int temp = arr[min];
					arr[min] = arr[j];
					arr[j] = temp;
				}
			}
			//the minimum index increases after each minimum value of the unsorted section is selected
			min++;
		}
		//giving the sorted array
		return arr;
	}

	//running selection sort on an array of integers
	public static void main(String[] args) {
		int[] arr = { 6, 3, 9, 1, 4, 0, 8, 2, 7, 5};
		System.out.println("Before sorting");
		for (int x : arr) {
			System.out.print(x + " ");
		}
		System.out.println("\nAfter sorting");
		for (int x : selection(arr)) {
			System.out.print(x + " ");
		}
	}

}

File: sort/selection_sort/ruby/selection_sort.rb
def selection_sort(arr)
  return arr if arr.length <= 1

  min_idx = nil

  (0..arr.length-2).each do |i|
    min_idx = i
    j = i + 1
    while j < arr.length
      if arr[j] < arr[min_idx]
        min_idx = j
      end
      j += 1
    end
    arr[i], arr[min_idx] = arr[min_idx], arr[i]
  end
  arr
end

arr = [6,5,3,1,8,7,2,4];
p selection_sort(arr);
